upload.sh

```diff
@@ -1,44 +1,102 @@
 #!/bin/sh
 
-# Get tag from environment
+# Get release tag from environment and checkout branch.
 tag=$(basename $GITHUB_REF)
+git checkout tags/$tag
 
-# Get action that triggered release event
+# Get package version.
+version=$(python setup.py --version)
+
+# Check if release tag matches the package version.
+pip install --quiet packaging==19.2
+
+match=$(python -c "
+from packaging.version import parse
+
+match = parse('$tag') == parse('$version')
+print(match)
+")
+
+if [ $match = "False" ]; then
+    echo "Release $tag does not match package $version"
+    exit 1
+fi
+
+# Get action that triggered event.
 action=$(python -c "
 import json
+
 with open('$GITHUB_EVENT_PATH', 'r') as file:
     event = json.load(file)
     action = event.get('action')
+
 print(action)
 ")
 
-echo "Release $tag was $action on GitHub ..."
+# Infer which repository to use.
+repository=$(python -c "
+from packaging.version import parse, Version
+
+version = parse('$tag')
+if isinstance(version, Version):
+    if version.is_devrelease:
+        print('Test PyPI')
 
-upload_to_pypi () {
-    # Checkout specified tag
-    git checkout tags/$tag
+    if not version.is_devrelease \
+        and not version.is_postrelease \
+        and not version.is_prerelease:
```

Comment:
Do we want to release post or pre releases to PyPI?

Reply:
I think it's more likely that we want to do a pre or post release to the real PyPI servers than a dev release, so we could reserve dev releases as the type that is automatically uploaded to the test server and the other are uploaded to the real server.

Reply:
Okay, updated script.